PYTHON_CODING.md

```diff
@@ -0,0 +1,216 @@
+
+   ```python
+def append(element, list_=[]):
+        list_.append(element)
+        return list_
+
+>>> print(append(0))
+[0]
+>>> print(append(1))
+[0, 1]
+```
+   instead prefer
+   ```python
+def append(element, list_=None):
+        if list_ is None:
+            list_ = []
+        list_.append(element)
+        return list_
+```
+- **Avoid generic type coercion when possible.** It is easy to catch in the mix unwanted types and mask a bug upstream.
+   ```python
+squared_lookup_table = [x*x for x in range(100)]
+
+def square_integers(integer_list):
+        """Element-wise square of the given list of integers."""
+        integer_list = [int(x) for x in integer_list]  # let's convert floats to int
+        return [squared_lookup_table[i] for i in integer_list]
+
+>>> square_integers([0, 4, 10])
+[0, 16, 100]
+>>> square_integers('92')  # you want this to fail to catch a bug!
+[81, 4]
+```
+   If you absolutely have to coerce a type, tighten it up to be sure you are coercing only those you want.
+   ```python
+def square_integers(integer_list):
+        integer_list = [int(round(x)) for x in integer_list]  # This always fails with strings.
+        return [squared_lookup_table[i] for i in integer_list]
+```
+   It is generally safer to leave any type of coercion to the user of the function since he knows what kind of data he intends to pass.
+- **Avoid abusing kwargs.** Using `**kwargs` is a general way to forward keyword arguments to another (possibly unknown) function, but when used instead of normal keyword arguments, it can mask bugs in your program.
+   ```python
+def decorate_str(decorated, **kwargs):
+        """Prepend and append a decoration string."""
+        separator = kwargs.get('separator', '_')  # default is '_'
+        decoration = kwargs.get('decoration_string', '')  # default is ''
+        return decoration + separator + decorated + separator + decoration
+
+>>> decorate_str('decorated', decoration_string='foo')
+'foo_decorated_foo'
+>>> decorate_str('decorated', decoration_str='foo')  # Whops! No Error because kwargs disrupt introspection!
+'_decorated_'
+```
+   If you want to use `**kwargs` instead of normal keyword argument, add manually some control logic.
+   ```python
+def decorate_str(decorated, **kwargs):
+        """Prepend and append a decoration string."""
+        separator = kwargs.pop('separator', '_')  # default is '_'
+        decoration = kwargs.pop('decoration_string', '')  # default is empty string
+        if kwargs != {}:
+            raise TypeError('got unexpected keyword arguments {}'.format(kwargs))
+        return decoration + separator + decorated + separator + decoration
+```
+   But it is usually more readable to just use keyword arguments (plus it does not break autocompletion).
+   ```python
+def decorate_str(decorated, separator='_', decoration_string=''):
+        return decoration_string + separator + decorated + separator + decoration_string
+```
+
+## Know the basics of OOP design
```

Comment:
Maybe "Object Oriented Programming (OOP) design principles will help you adopt a design that will grow with your project"?